File: packages/plugin-codegen/src/options.ts
import { z } from 'zod'
import { defineFlags } from '@chkit/core'

import { CodegenConfigError } from './errors.js'

// ───── Plugin config schema (what codegen({...}) accepts) ─────

export const PluginConfigSchema = z.object({
  outFile: z.string().min(1).optional(),
  emitZod: z.boolean().optional(),
  tableNameStyle: z.enum(['pascal', 'camel', 'raw']).optional(),
  bigintMode: z.enum(['string', 'bigint']).optional(),
  includeViews: z.boolean().optional(),
  runOnGenerate: z.boolean().optional(),
  failOnUnsupportedType: z.boolean().optional(),
  emitIngest: z.boolean().optional(),
  ingestOutFile: z.string().min(1).optional(),
  emitMigrations: z.boolean().optional(),
  migrationsOutFile: z.string().min(1).optional(),
})

export type PluginConfig = z.input<typeof PluginConfigSchema>

// ───── Codegen command schema ─────

export const CodegenSchema = z.object({
  outFile: z.string().min(1).default('./src/generated/chkit-types.ts'),
  emitZod: z.boolean().default(false),
  tableNameStyle: z.enum(['pascal', 'camel', 'raw']).default('pascal'),
  bigintMode: z.enum(['string', 'bigint']).default('string'),
  includeViews: z.boolean().default(false),
  runOnGenerate: z.boolean().default(true),
  failOnUnsupportedType: z.boolean().default(true),
  emitIngest: z.boolean().default(false),
  ingestOutFile: z.string().min(1).default('./src/generated/chkit-ingest.ts'),
  emitMigrations: z.boolean().default(false),
  migrationsOutFile: z.string().min(1).default('./src/generated/chkit-migrations.ts'),
})
export type CodegenOptions = z.infer<typeof CodegenSchema>

// ───── CLI flag definitions ─────

export const CODEGEN_FLAGS = defineFlags([
  { name: '--check', type: 'boolean', description: 'Check if generated output is up-to-date' },
  { name: '--out-file', type: 'string', description: 'Output file path', placeholder: '<path>' },
  { name: '--emit-zod', type: 'boolean', description: 'Emit Zod schemas alongside TypeScript types', negation: true },
  { name: '--emit-ingest', type: 'boolean', description: 'Emit ingest helper functions', negation: true },
  { name: '--ingest-out-file', type: 'string', description: 'Ingest output file path', placeholder: '<path>' },
  { name: '--bigint-mode', type: 'string', description: 'How to represent large integers (string or bigint)', placeholder: '<mode>' },
  { name: '--include-views', type: 'boolean', description: 'Include views in generated output' },
  { name: '--emit-migrations', type: 'boolean', description: 'Emit runtime migration module', negation: true },
  { name: '--migrations-out-file', type: 'string', description: 'Migration module output file path', placeholder: '<path>' },
] as const)

// ───── Flag mappings ─────

interface FlagMappingEntry {
  key: string
  coerce?: (value: string) => unknown
}

type FlagMapping = Record<string, FlagMappingEntry>

const CODEGEN_FLAG_MAP: FlagMapping = {
  '--out-file': { key: 'outFile' },
  '--emit-zod': { key: 'emitZod' },
  '--emit-ingest': { key: 'emitIngest' },
  '--ingest-out-file': { key: 'ingestOutFile' },
  '--bigint-mode': { key: 'bigintMode' },
  '--include-views': { key: 'includeViews' },
  '--emit-migrations': { key: 'emitMigrations' },
  '--migrations-out-file': { key: 'migrationsOutFile' },
}

// ───── mapFlags ─────

function mapFlags(
  flags: Record<string, string | string[] | boolean | undefined>,
  mapping: FlagMapping
): Record<string, unknown> {
  const result: Record<string, unknown> = {}
  for (const [flagName, entry] of Object.entries(mapping)) {
    const raw = flags[flagName]
    if (raw === undefined) continue
    if (entry.coerce && typeof raw === 'string') {
      result[entry.key] = entry.coerce(raw)
    } else {
      result[entry.key] = raw
    }
  }
  return result
}

// ───── resolveOptions ─────

function resolveOptions<T extends z.ZodTypeAny>(
  schema: T,
  pluginConfig: Record<string, unknown>,
  runtimeOptions: Record<string, unknown>,
  flags: Record<string, string | string[] | boolean | undefined>,
  flagMapping: FlagMapping
): z.infer<T> {
  const cliOverrides = mapFlags(flags, flagMapping)
  const merged = { ...pluginConfig, ...runtimeOptions, ...cliOverrides }
  const result = schema.safeParse(merged)
  if (!result.success) {
    const issue = result.error.issues[0]
    throw new CodegenConfigError(
      issue ? `${issue.path.join('.')}: ${issue.message}` : 'Invalid codegen options'
    )
  }
  return result.data
}

export function resolveCodegenOptions(
  pluginConfig: Record<string, unknown>,
  runtimeOptions: Record<string, unknown>,
  flags: Record<string, string | string[] | boolean | undefined>
): CodegenOptions {
  return resolveOptions(CodegenSchema, pluginConfig, runtimeOptions, flags, CODEGEN_FLAG_MAP)
}

/** Fill defaults for partial options. Used by generators. */
export function normalizeCodegenOptions(options: PluginConfig = {}): CodegenOptions {
  return CodegenSchema.parse(options)
}

export function isRunOnGenerateEnabled(
  pluginConfig: Record<string, unknown>,
  runtimeOptions: Record<string, unknown>
): boolean {
  return resolveOptions(CodegenSchema, pluginConfig, runtimeOptions, {}, {}).runOnGenerate
}

File: chkit_python/src/chkit_plugin_codegen/options.py
"""Codegen plugin options (Pydantic-validated) + CLI flag definitions."""

from __future__ import annotations

from typing import Any, Literal, TypeAlias

from pydantic import BaseModel, ConfigDict, Field, ValidationError, field_validator

from chkit_plugin_codegen.errors import CodegenConfigError

TableNameStyle: TypeAlias = Literal["pascal", "camel", "raw"]
BigIntMode: TypeAlias = Literal["str", "int"]

# TS uses ``'string' | 'bigint'``; Python uses ``'str' | 'int'`` (Python's int
# is unbounded so it's a strict superset of TS bigint — no precision loss).
# Accept the TS spellings for cross-language config portability and normalize
# to the Python form.
_BIGINT_TS_ALIASES: dict[str, BigIntMode] = {
    "string": "str",
    "bigint": "int",
}


class CodegenOptions(BaseModel):
    """Fully-resolved options used by the generator (all defaults filled in)."""

    out_file: str = Field(default="./src/generated/chkit_models.py", alias="outFile")
    table_name_style: TableNameStyle = Field(default="pascal", alias="tableNameStyle")
    bigint_mode: BigIntMode = Field(default="int", alias="bigintMode")
    include_views: bool = Field(default=False, alias="includeViews")
    run_on_generate: bool = Field(default=True, alias="runOnGenerate")
    fail_on_unsupported_type: bool = Field(
        default=True, alias="failOnUnsupportedType"
    )

    model_config = ConfigDict(
        frozen=True,
        extra="forbid",
        populate_by_name=True,
    )

    @field_validator("bigint_mode", mode="before")
    @classmethod
    def _coerce_bigint_mode(cls, value: object) -> object:
        if isinstance(value, str) and value in _BIGINT_TS_ALIASES:
            return _BIGINT_TS_ALIASES[value]
        return value


class PluginConfig(BaseModel):
    """User-supplied options to ``codegen({...})``. All fields are optional."""

    out_file: str | None = Field(default=None, alias="outFile")
    table_name_style: TableNameStyle | None = Field(default=None, alias="tableNameStyle")
    bigint_mode: BigIntMode | None = Field(default=None, alias="bigintMode")
    include_views: bool | None = Field(default=None, alias="includeViews")
    run_on_generate: bool | None = Field(default=None, alias="runOnGenerate")
    fail_on_unsupported_type: bool | None = Field(
        default=None, alias="failOnUnsupportedType"
    )

    model_config = ConfigDict(
        frozen=True,
        extra="forbid",
        populate_by_name=True,
    )

    @field_validator("bigint_mode", mode="before")
    @classmethod
    def _coerce_bigint_mode(cls, value: object) -> object:
        if isinstance(value, str) and value in _BIGINT_TS_ALIASES:
            return _BIGINT_TS_ALIASES[value]
        return value


def normalize_codegen_options(
    options: PluginConfig | CodegenOptions | dict[str, Any] | None,
) -> CodegenOptions:
    """Fill in defaults for any missing fields and return a fully-resolved ``CodegenOptions``."""
    if options is None:
        return CodegenOptions()
    if isinstance(options, CodegenOptions):
        return options
    if isinstance(options, PluginConfig):
        payload = options.model_dump(exclude_none=True, by_alias=False)
    else:
        try:
            payload = PluginConfig.model_validate(options).model_dump(
                exclude_none=True, by_alias=False
            )
        except ValidationError as error:
            raise CodegenConfigError(str(error)) from error
    try:
        return CodegenOptions.model_validate(payload)
    except ValidationError as error:
        raise CodegenConfigError(str(error)) from error


# CLI flag definitions — mirror the TS ``defineFlags`` shape, just as dicts so
# the existing plugin runtime can introspect them.
CODEGEN_FLAGS: list[dict[str, Any]] = [
    {
        "name": "--check",
        "type": "boolean",
        "description": "Check if generated output is up-to-date",
    },
    {
        "name": "--out-file",
        "type": "string",
        "description": "Output file path",
        "placeholder": "<path>",
    },
    {
        "name": "--bigint-mode",
        "type": "string",
        "description": "How to represent 64-bit integers (int or str)",
        "placeholder": "<mode>",
    },
    {
        "name": "--include-views",
        "type": "boolean",
        "description": "Include views in generated output",
    },
    {
        "name": "--table-name-style",
        "type": "string",
        "description": "Class naming convention (pascal / camel / raw)",
        "placeholder": "<style>",
    },
]

# Flag → CodegenOptions field mapping (used by the runtime when promoting flags
# into the ``options`` dict).
CODEGEN_FLAG_MAP: dict[str, dict[str, str]] = {
    "--out-file": {"key": "out_file"},
    "--bigint-mode": {"key": "bigint_mode"},
    "--include-views": {"key": "include_views"},
    "--table-name-style": {"key": "table_name_style"},
}


__all__ = [
    "CODEGEN_FLAGS",
    "CODEGEN_FLAG_MAP",
    "BigIntMode",
    "CodegenOptions",
    "PluginConfig",
    "TableNameStyle",
    "normalize_codegen_options",
]
